modules/auth.php

```diff
@@ -260,6 +262,92 @@ function handler_admin_authgroupesx($page, $action = 'list', $id = null)
         $table_editor->describe('last_used', 'dernière utilisation', true);
         $table_editor->apply($page, $action, $id);
     }
+
+    /** Handles the Discourse SSO authentication.
+     * https://meta.discourse.org/t/official-single-sign-on-for-discourse/13045
+     * Expects the Discourse domain as the last component of the URL path
+     * Expects the following GET parameters: sso, sig
+     * e.g. https://www.polytechnique.org/auth-discourse/forum.polytechnique.org?sso=...&sig=...
+     */
+    function handler_discourse($page, $domain = '')
+    {
+        global $globals;
+
+        // Load the key
+        if (!preg_match('/^[a-zA-Z0-9\-.:]+$/', $domain)) {
+            $page->kill("Domaine non valide");
+        } elseif (empty($globals->discourse->$domain)) {
+            $page->kill("Domaine inconnu");
+        }
+        $ext_url = 'http://' . $domain. '/session/sso_login';
+        $key = $globals->discourse->$domain;
+
+        // Chech the signature of the given nonce
+        $sso_data_b64 = Get::s('sso');
+        $sign = hash_hmac('sha256', $sso_data_b64, $key);
+        if (!secure_string_compare($sign, Get::s('sig'))) {
+            $page->kill("Signature invalide");
+        }
+        $sso_data = array();
+        parse_str(base64_decode($sso_data_b64), $sso_data);
+        if (empty($sso_data['nonce'])) {
+            $page->kill("Données SSO non valides");
+        }
+        if (!empty($sso_data['return_sso_url']) && $sso_data['return_sso_url'] !== $ext_url) {
```

Comment:
``return_sso_url`` field is not "officially" documented in https://meta.discourse.org/t/official-single-sign-on-for-discourse/13045 but is present in the version of Discourse I've tested (it was a local install), which is why I used ``&&``. If we're only supporting Discourse versions which use ``return_sso_url``, I agree this ``if`` should use ``||`` instead.